### ISSUE TITLE
Reinforcement of OOB types for Next.js and Typescript. 

### ISSUE THREAD

**Issue by Maxbrain0** [inland-empire-software-development/spark#30]
**Components that require updating:**
- Next.js typings

**Tech to be used:**
- Next.js
- Typescript

**What is the current behavior:**
We explicitly importing types from index.d.ts to avoid name clashes (eg, Request is also a built in type for the built-in fetch API). Try to use out of the box types as much as possible without sacrificing type support for sake of making linter happy. 

**What is the new behavior:**
- [ ] Type the user context used throughout the app. 
- [ ] Type the _app React class-based component so the state is typed.
- [ ] Use built in typescript types for nextjs page components (whether functional or class-based) and API request and responses. [Next documentation here](https://nextjs.org/docs/basic-features/typescript). 

**Justification:** Typing the user state in the context and provider that wraps the app is most important as people will get better IDE support when accessing the user inside of their components. 

Notes: The API and pages could be updated incrementally without any problems

**Comment by lloan** [comment]
@Maxbrain0 great suggestion, this has been groomed and updated. Can you take a look at this? Create a new branch off of the master branch for this enhancement and push up to dev tier as you test. I think this has a high priority as it would make working on the rest of the project much easier. 

Save what you do to the project here and if you have time and are interested in submitting a PR for next-authenticate, that would be super helpful as we will be basing future projects off of it as well. 

[Next-authenticate issue for this](https://github.com/lloan/next-authenticate/issues/8)

**Comment by Maxbrain0** [comment]
@lloan, 

Alright, I already played around with this yesterday. I may touch base with you or @tonynguyen111997.

Which one of you mainly handled the logic inside of _app? I ask because there was some logic where state was first null, but then became either a boolean or string for handling redirects. What I have currently seems to work, but I worry I introduced a bug. 

Also going to look into branch naming conventions.

**Comment by Maxbrain0** [comment]
Hey, I didn't realize I'd posted conversation in the [next-authenticate issue](https://github.com/lloan/next-authenticate/issues/8). We have IDE support in my current state. My only question is which properties does the context need? You were providing the whole state previously. Do we need _isPublic_ and _redirect_ in the pages? Anyhow, it's easy enough to tweak in the future. 

Also, any opinions on naming conventions, I'm open too. I call the type MyAppContext because it's the parent context in MyApp, but we could just as well call is User, or UserAuth, etc.

```
import {createContext} from 'react';
import {MyAppContext} from '../';


const defaultAuth = {
  user: undefined,
  access: false,
  redirect: undefined,
  isPublic: false,
};

```



const Context = createContext<MyAppContext>(defaultAuth);


export default Context;

![Screen Shot 2020-01-15 at 4 22 49 PM](https://user-images.githubusercontent.com/8978605/72482780-b5faaf00-37b3-11ea-945c-aa10feddf38f.png)

**Comment by lloan** [comment]
So all the pages should have access to the store and be able to only consume what they need. With what you did, it should be easier for team members to see what hey have access to, so that's wonderful. Each page will definitely use `user, isPublic, redirect and access` - we will be using this to provide data to the components, but can' foresee what other items we will input in there yet.

**Comment by lloan** [comment]
This was merged in #46 - pushed to all tiers.
